pkg/pullexecutor/pullexecutor.go

```diff
@@ -0,0 +1,147 @@
+	PullSecrets map[string]string
+	Image       string
+}
+
+// PullExecutor executes the pulls
+type PullExecutor struct {
+	asyncPull      bool
+	imageSvcClient cri.ImageServiceClient
+	mutex          *sync.Mutex
+	asyncErrs      map[docker.Named]error
+	secretStore    secret.Store
+	mounter        *backend.SnapshotMounter
+}
+
+// NewPullExecutor initializes a new pull executor object
+func NewPullExecutor(o *PullExecutorOptions) *PullExecutor {
+	return &PullExecutor{
+		asyncPull:      o.AsyncPull,
+		mutex:          &sync.Mutex{},
+		imageSvcClient: o.ImageServiceClient,
+		secretStore:    o.SecretStore,
+	}
+}
+
+// StartPulling starts pulling the image
+func (m *PullExecutor) StartPulling(o *PullOptions) error {
+
+	keyring, err := m.secretStore.GetDockerKeyring(o.Context, o.PullSecrets)
+	if err != nil {
+		return errors.Errorf("unable to fetch keyring: %s", err)
+	}
+
+	if !m.asyncPull {
+		puller := remoteimage.NewPuller(m.imageSvcClient, o.NamedRef, keyring)
+		shouldPull := o.PullAlways || !m.mounter.ImageExists(o.Context, o.NamedRef)
+		if shouldPull {
+			klog.Infof("pull image %q ", o.Image)
+			if err = puller.Pull(o.Context); err != nil {
+				pullstatus.Update(o.NamedRef, pullstatus.Errored)
+				return errors.Errorf("unable to pull image %q: %s", o.Image, err)
+			}
+		}
+		return nil
+	}
+
+	if pullstatus.Get(o.NamedRef) == pullstatus.Pulled ||
+		pullstatus.Get(o.NamedRef) == pullstatus.StillPulling {
+		return nil
+	}
+
+	go func() {
+		if pullstatus.Get(o.NamedRef) == pullstatus.StatusNotFound {
+			m.mutex.Lock()
+			c, cancel := context.WithTimeout(context.Background(), pullCtxTimeout)
+			defer cancel()
+
+			if pullstatus.Get(o.NamedRef) == pullstatus.StillPulling {
+				m.mutex.Unlock()
+				return
+			}
+
+			puller := remoteimage.NewPuller(m.imageSvcClient, o.NamedRef, keyring)
+			shouldPull := o.PullAlways || !m.mounter.ImageExists(o.Context, o.NamedRef)
+			if shouldPull {
+				klog.Infof("pull image %q ", o.Image)
+				pullstatus.Update(o.NamedRef, pullstatus.StillPulling)
+				if err = puller.Pull(c); err != nil {
+					pullstatus.Update(o.NamedRef, pullstatus.Errored)
+					m.asyncErrs[o.NamedRef] = fmt.Errorf("unable to pull image %q: %s", o.Image, err)
+					m.mutex.Unlock()
+					return
+				}
+				pullstatus.Update(o.NamedRef, pullstatus.Pulled)
+
+			}
+			m.mutex.Unlock()
+		}
+	}()
+
+	return nil
+}
+
+// WaitForPull waits until the image pull succeeds or errors or timeout is exceeded
+func (m *PullExecutor) WaitForPull(o *PullOptions) error {
+	if !m.asyncPull {
+		return nil
+	}
+
+	condFn := func() (done bool, err error) {
+		if pullstatus.Get(o.NamedRef) == pullstatus.Pulled {
+			return true, nil
+		}
+
+		if m.asyncErrs[o.NamedRef] != nil {
+			return false, m.asyncErrs[o.NamedRef]
+		}
+		return false, nil
+	}
+
+	if err := wait.PollImmediate(
+		pullPollTimeInterval,
+		pullPollTimeout,
+		condFn); err != nil {
+		return errors.Errorf("waited too long to download the image: %v", err)
+	}
```

Comment:
channel, channels :)

Reply:
Not a big fan of channels (unless there is no other option or benefits outweigh the cons). They make the code harder to read.

Reply:
We can take this in a separate issue if you have a differing opinion.